api/core/model_runtime/model_providers/yi/llm/llm.py

```diff
@@ -0,0 +1,988 @@
+
+        # every reply is primed with <im_start>assistant
+        num_tokens += 3
+
+        if tools:
+            num_tokens += self._num_tokens_for_tools(encoding, tools)
+
+        return num_tokens
+
+    def _num_tokens_for_tools(self, encoding: tiktoken.Encoding, tools: list[PromptMessageTool]) -> int:
+        """
+        Calculate num tokens for tool calling with tiktoken package.
+
+        :param encoding: encoding
+        :param tools: tools for tool calling
+        :return: number of tokens
+        """
+        num_tokens = 0
+        for tool in tools:
+            num_tokens += len(encoding.encode('type'))
+            num_tokens += len(encoding.encode('function'))
+
+            # calculate num tokens for function object
+            num_tokens += len(encoding.encode('name'))
+            num_tokens += len(encoding.encode(tool.name))
+            num_tokens += len(encoding.encode('description'))
+            num_tokens += len(encoding.encode(tool.description))
+            parameters = tool.parameters
+            num_tokens += len(encoding.encode('parameters'))
+            if 'title' in parameters:
+                num_tokens += len(encoding.encode('title'))
+                num_tokens += len(encoding.encode(parameters.get("title")))
+            num_tokens += len(encoding.encode('type'))
+            num_tokens += len(encoding.encode(parameters.get("type")))
+            if 'properties' in parameters:
+                num_tokens += len(encoding.encode('properties'))
+                for key, value in parameters.get('properties').items():
+                    num_tokens += len(encoding.encode(key))
+                    for field_key, field_value in value.items():
+                        num_tokens += len(encoding.encode(field_key))
+                        if field_key == 'enum':
+                            for enum_field in field_value:
+                                num_tokens += 3
+                                num_tokens += len(encoding.encode(enum_field))
+                        else:
+                            num_tokens += len(encoding.encode(field_key))
+                            num_tokens += len(encoding.encode(str(field_value)))
+            if 'required' in parameters:
+                num_tokens += len(encoding.encode('required'))
+                for required_field in parameters['required']:
+                    num_tokens += 3
+                    num_tokens += len(encoding.encode(required_field))
+
+        return num_tokens
+
+    def get_customizable_model_schema(self, model: str, credentials: dict) -> AIModelEntity:
+        """
+            OpenAI supports fine-tuning of their models. This method returns the schema of the base model
+            but renamed to the fine-tuned model name.
+
+            :param model: model name
+            :param credentials: credentials
+
+            :return: model schema
+        """
+        if not model.startswith('ft:'):
```

Comment:
Does Yi support fine-tuning models? This is the specific logic for fine-tuning models.

Reply:
There probably isn't one at the moment, but I'm not sure if it will be fully compatible with OpenAI's gameplay in the future. Maybe I should remove it first?